### PR TITLE
[ADD] pq_history

### PR DESCRIPTION
**Pull request by kongrattapong**

### CODE REVIEW

File: sale_prequotation/sale_prequotation.py [hunk 1]
```diff
@@ -869,6 +873,78 @@ def action_sale(self, cr, uid, ids, format_quote, context=None):
...
+                'labour_sale_id': prequote.labour_sale_id.id,
+                'mat_sale_no': prequote.mat_sale_no,
+                'labour_sale_no': prequote.labour_sale_no,
+                'doc_version': prequote.doc_version,
+                'pq_version': prequote.pq_version,
+                'pq_id': prequote.id,
+                'customer_commission': prequote.customer_commission,
+                'percent_unforeseen': prequote.percent_unforeseen,
+            }
+            pq_older = pq.create(cr, uid, vals, context=context)
```
Comment (Saran440) [inline sale_prequotation/sale_prequotation.py]:
```suggestion
            pq_new = pq.create(cr, uid, vals, context=context)
```
Should be use `pq_new` and `pq_origin` are good?

File: sale_prequotation/sale_prequotation.py [hunk 2]
```diff
@@ -869,6 +873,78 @@ def action_sale(self, cr, uid, ids, format_quote, context=None):
...
+                    'percentage': lines.percentage,
+                    'cost_price_unit': lines.cost_price_unit,
+                    'discount': lines.discount,
+                    'profit': lines.profit,
+                    'insurance': lines.insurance,
+                    'transport': lines.transport,
+                }
+                pq_product.create(cr, uid, vals, context=context)
+            pq.write(cr, uid, [pq_older], {'state': 'cancel'}, context)
+            self.write(cr, uid, ids, {
```
Comment (Saran440) [inline sale_prequotation/sale_prequotation.py]: prequote.pq_version += 1

File: sale_prequotation/sale_prequotation.py [hunk 3]
```diff
@@ -869,6 +873,78 @@ def action_sale(self, cr, uid, ids, format_quote, context=None):
...
+                'mat_sale_no': prequote.mat_sale_no,
+                'labour_sale_no': prequote.labour_sale_no,
+                'doc_version': prequote.doc_version,
+                'pq_version': prequote.pq_version,
+                'pq_id': prequote.id,
+                'customer_commission': prequote.customer_commission,
+                'percent_unforeseen': prequote.percent_unforeseen,
+            }
+            pq_older = pq.create(cr, uid, vals, context=context)
+            for lines in prequote.order_line_labour:
```
Comment (Saran440) [inline sale_prequotation/sale_prequotation.py]:
```suggestion
            vals = {}
            for lines in prequote.order_line_labour:
                vals += {
                   ...
                }
            pq_labour.create(cr, uid, vals, context=context)
```

**Review by Saran440 — COMMENTED** [review]
Review code

File: sale_prequotation/sale_prequotation.py [hunk 4]
```diff
@@ -298,10 +299,11 @@ def _get_currency(self, cr, uid, context=None):
             ], 'Status', readonly=True,
              select=True, track_visibility='always',),
         'partner_id': fields.many2one('res.partner', 'Customer', domain=[('customer', '=', True), ('is_company', '=', True)], required=False, readonly=True,
-                                      states={'draft': [('readonly', False)], 'confirm': [('readonly', False), ('required', True)]}, 
+                                      states={'draft': [('readonly', False)], 'confirm': [('readonly', False), ('required', True)]},
                                       select=True, track_visibility='always'),
         #'partner_contact_id': fields.char('Contact Person Customer', size=128, required=False, states={'confirm': [('required', True)]}),
         'sale_ids': fields.one2many('sale.order', 'pq_id', 'Calculation Sheet', readonly=True),
+        'pq_history_line': fields.one2many('sale.prequotation', 'pq_id', 'PQ History'),
```
Comment (kittiu) [inline sale_prequotation/sale_prequotation.py]:
```suggestion
        'pq_history_ids': fields.one2many('sale.prequotation', 'pq_id', 'PQ History'),
```

File: sale_prequotation/sale_prequotation.py [hunk 5]
```diff
@@ -869,6 +873,89 @@ def action_sale(self, cr, uid, ids, format_quote, context=None):
         result['domain'] = "[('id','in', [" + ','.join(map(str, quote_ids)) + "])]"
         return result
 
+    def _prepare_pq(self, cr, uid, id, prequote, context=None):
```
Comment (kittiu) [inline sale_prequotation/sale_prequotation.py]: Do not created this fuctions, please use copy()

File: sale_prequotation/sale_prequotation.py [hunk 6]
```diff
@@ -869,6 +874,25 @@ def action_sale(self, cr, uid, ids, format_quote, context=None):
         result['domain'] = "[('id','in', [" + ','.join(map(str, quote_ids)) + "])]"
         return result
 
+    def action_new_pq_version(self, cr, uid, ids, context=None):
+        for prequote in self.browse(cr, uid, ids, context=context):
+            pq_new = self.copy(cr, uid, prequote.id, context=context)
+            self.write(cr, uid, self.search(cr, uid, [('id', '=', pq_new)]), {
```
Comment (kittiu) [inline sale_prequotation/sale_prequotation.py]: why not just self.write(cr, uid, [new_pq_id], context=context)

File: sale_prequotation/sale_prequotation.py [hunk 7]
```diff
@@ -869,6 +874,25 @@ def action_sale(self, cr, uid, ids, format_quote, context=None):
...
+                'name': prequote.name + '-' + NUMBER_TO_CHAR[prequote.pq_version],
+                'pq_id': prequote.id,
+                'date_pq': prequote.date_pq,
```
Comment (kittiu) [inline sale_prequotation/sale_prequotation.py:883]: Why need so many update?

File: sale_prequotation/sale_prequotation.py [hunk 8]
```diff
@@ -869,6 +874,25 @@ def action_sale(self, cr, uid, ids, format_quote, context=None):
         result['domain'] = "[('id','in', [" + ','.join(map(str, quote_ids)) + "])]"
         return result
 
+    def action_new_pq_version(self, cr, uid, ids, context=None):
+        for prequote in self.browse(cr, uid, ids, context=context):
+            pq_new = self.copy(cr, uid, prequote.id, context=context)
```
Comment (kittiu) [inline sale_prequotation/sale_prequotation.py]:
```suggestion
            new_pq_id = self.copy(cr, uid, prequote.id, context=context)
```

File: sale_prequotation/sale_prequotation.py (hunk 6 above)
Comment (kongrattapong) [reply sale_prequotation/sale_prequotation.py]: Fixed it

File: sale_prequotation/sale_prequotation.py (hunk 7 above)
Comment (kongrattapong) [reply sale_prequotation/sale_prequotation.py:883]: cuz in def copy(

values is set to this:
```default.update({
            'date_pq': time.strftime('%Y-%m-%d'),
            'cal_date': time.strftime('%Y-%m-%d'),
            'state': 'draft',
            'user_id': uid,
            'sale_ids': False,
            'mat_sale_id': None,
            'labour_sale_id': None,
            'doc_version': 0,
            'mat_sale_no': False,
            'labour_sale_no': False,
            'pq_history_ids': False,
            'name': self.pool.get('ir.sequence').get(cr, uid, 'sale.prequotation') or '/',
        })```

**Pull request merged**